File: ios/Sources/NearbyPlugin/Classes/Events/PayloadTransferUpdateEvent.swift
import Foundation
import Capacitor

import NearbyConnections

@objc public class PayloadTransferUpdateEvent: EndpointEvent {
    let payloadID: PayloadID

    let update: TransferUpdate

    init(_ endpointID: EndpointID, _ payloadID: PayloadID, update: TransferUpdate) {
        self.payloadID = payloadID

        self.update = update

        super.init(endpointID)
    }

    override public func toJSObject() -> JSObject {
        var result = super.toJSObject()

        result["payloadID"] = String(payloadID)

        result["status"] = update.toString()

        switch update {
        case .success, .canceled, .failure:
            break
        case .progress(let progress):
            result["bytesTransferred"] = String(progress.completedUnitCount)
            result["totalBytes"] = String(progress.totalUnitCount)
        }

        return result
    }
}

File: android/src/main/java/com/getcapacitor/community/classes/events/PayloadTransferUpdateEvent.java
package com.getcapacitor.community.classes.events;

import androidx.annotation.NonNull;
import com.getcapacitor.JSObject;
import com.getcapacitor.community.classes.Endpoint;
import com.getcapacitor.community.classes.PayloadTransferUpdate;

public class PayloadTransferUpdateEvent extends EndpointEvent {

    @NonNull
    PayloadTransferUpdate update;

    public PayloadTransferUpdateEvent(@NonNull Endpoint endpoint, @NonNull PayloadTransferUpdate update) {
        super(endpoint);
        this.update = update;
    }

    @NonNull
    public JSObject toJSObject() {
        JSObject result = super.toJSObject();

        result.put("payloadID", update.payloadID());

        result.put("status", update.getStatus());

        result.put("bytesTransferred", update.bytesTransferred());
        result.put("totalBytes", update.totalBytes());

        return result;
    }
}
